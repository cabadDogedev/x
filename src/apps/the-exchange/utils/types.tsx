import { BridgingProvider, ExchangeOffer } from '@etherspot/prime-sdk/dist/sdk/data';
import { Route } from '@lifi/sdk';

export enum CardPosition {
    SWAP = 'SWAP',
    RECEIVE = 'RECEIVE'
}

export type SwapType = {
    fromAmount: number,
    fromTokenAddress: string,
    fromChainId: number;
    fromTokenDecimals: number;
    toTokenAddress: string,
    toChainId: number,
    toTokenDecimals: number;
    slippage: number;
    fromAccountAddress?: string
    provider?: BridgingProvider;
}

export type SwapOffer = {
    tokenAmountToReceive: number;
    offer: Route | ExchangeOffer;
}

export type AmountType = {
    tokenAmount: number;
    usdAmount: number;
  }
  
export type ChainType = {
    chainId: number;
    chainName: string;
  }
