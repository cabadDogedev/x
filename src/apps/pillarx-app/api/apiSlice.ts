import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { ApiResponse } from '../../../types/api'

export const pillarXApi = createApi({
    reducerPath: 'pillarXApi',
    baseQuery: fetchBaseQuery({ baseUrl: 'https://feed-nubpgwxpiq-uc.a.run.app/' }),
    endpoints: (builder) => ({
        getTilesInfo: builder.query<ApiResponse, { page: number; address: string }>({
            query: ({ page, address }) => `?page=${page}&address=${address}`,
        }),
    })
})

export const { useGetTilesInfoQuery } = pillarXApi;
