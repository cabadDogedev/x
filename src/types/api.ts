export type Asset = {
    name: string;
    symbol: string;
    id: number;
    contracts: string[];
    logo: string;
    blockchains: string[];
  }
  
  export type CrossChainBalance = {
    balance: number;
    balanceRaw: string;
    chainId: string;
    address: string;
  }
  
  export type ContractBalance = {
    balance: number;
    balanceRaw: string;
    chainId: string;
    address: string;
    decimals: number;
  }
  
  export type AssetData = {
    asset: Asset;
    realized_pnl: number;
    unrealized_pnl: number;
    allocation: number;
    price: number;
    price_bought: number;
    price_change_24h: number;
    price_change_1h: number;
    total_invested: number;
    min_buy_price: number;
    max_buy_price: number;
    estimated_balance: number;
    token_balance: number;
    cross_chain_balances: {
      [key: string]: CrossChainBalance;
    };
    contracts_balances: ContractBalance[];
  }
  
  export type TotalPnlHistory = {
    realized: number;
    unrealized: number;
  }
  
  export type ApiData = {
    total_wallet_balance: number;
    wallet: string;
    wallets: string[];
    total_realized_pnl: number;
    total_unrealized_pnl: number;
    assets: AssetData[];
    total_pnl_history: {
      [key in '24h' | '7d' | '30d' | '1y']?: TotalPnlHistory;
  };
  
  }
  
  // TO DO - meta type to change when api ready, layout should be enum
  export type Projection = {
    meta: unknown;
    data: ApiData;
    layout: string;
    id: string;
  }
  
  export type ApiResponse = {
    projection: Projection[];
  }
  