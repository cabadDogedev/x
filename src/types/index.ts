export interface AppManifest {
  title: string;
  description: string;
  translations: Record<string, string>;
}
