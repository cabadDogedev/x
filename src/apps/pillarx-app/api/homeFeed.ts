/* eslint-disable no-console */
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { addMiddleware } from '../../../store';
import { ApiResponse } from '../../../types/api';

export const homeFeedApi = createApi({
    reducerPath: 'homeFeedApi',
    baseQuery: fetchBaseQuery({ baseUrl:
        process.env.REACT_APP_USE_TESTNETS === 'true' ?
        'https://feed-nubpgwxpiq-uc.a.run.app' :
        'https://feed-7eu4izffpa-uc.a.run.app'
    }),
    endpoints: (builder) => ({
        getTilesInfo: builder.query<ApiResponse, { page: number; address: string }>({
            query: ({ page, address }) => `?page=${page}&address=${address}`,
        }),
    })
})

/**
 * Add this to the store
 */
addMiddleware(homeFeedApi);

export const { useGetTilesInfoQuery } = homeFeedApi;
