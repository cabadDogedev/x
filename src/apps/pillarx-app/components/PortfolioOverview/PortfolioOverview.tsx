import { useTranslation } from 'react-i18next'

// types
import { Projection, WalletPortfolioData } from '../../../../types/api'

// images
import DefaultLogo from '../../images/logo-unknown.png'
import BlendIcon from '../../images/blend-icon.svg'
import RoutingIcon from '../../images/routing-icon.svg'

// components
import Tags from '../Tags/Tags'
import TileContainer from '../TileContainer/TileContainer'
import Body from '../Typography/Body'
import H1 from '../Typography/H1'
import WalletAddressOverview from '../WalletAdddressOverview/WalletAddressOverview'
import TokensHorizontalList from '../TokensHorizontalList/TokensHorizontalList'
import TokensPercentage from '../TokensPercentage/TokensPercentage'
import SkeletonLoader from '../../../../components/SkeletonLoader'
import { getAllUniqueBlockchains } from '../../utils/blockchain'


type PortfolioOverviewProps = {
    data: Projection | undefined;
    isDataLoading: boolean;
}

const PortfolioOverview = ({ data, isDataLoading }: PortfolioOverviewProps) => {
    const [t] = useTranslation();

    const { data: dataPortlioOverview } = data || {};

    const { assets = [], total_pnl_history = {}, total_wallet_balance = 0, wallet = '' } = dataPortlioOverview as WalletPortfolioData || {};
    const { realized: pnl24hRealized = 0, unrealized: pnl24hUnrealized = 0 } = total_pnl_history['24h'] || {};

    const numberOfTokens = assets.length ?? 0;
    
    const allBlockchains = assets.map((asset) => asset.asset.blockchains).flat();

    const allBlockchainsLogos = assets.map((asset) => asset.asset.logo ? asset.asset.logo : DefaultLogo).flat();

    const numberOfBlockchains = getAllUniqueBlockchains(allBlockchains).length ?? 0;

    const totalPnl24h = pnl24hRealized + pnl24hUnrealized;
    
    const percentageChange = (totalPnl24h / (total_wallet_balance)) * 100;

if (!data || isDataLoading) {
    return (
    <TileContainer className='p-10 gap-20 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23] mobile:flex-col mobile:gap-4'>
        <div className='flex flex-col justify-between'>
            <SkeletonLoader $height='40px' $width='200px' $marginBottom='54px' $radius='10px' className='mobile:hidden' />
        <div className='mobile:border mobile:border-[#312F3A] mobile:rounded-[10px] mobile:p-4 mobile:w-full'>
            <SkeletonLoader $height='77px' $width='180px' $radius='10px' />
        </div>
        </div>
            <div className='flex w-full flex-col items-end justify-end mobile:justify-normal gap-5'>
                <div className='flex w-full gap-4 justify-end mobile:flex-row tablet:flex-col tablet:items-end mobile:flex-row mobile:justify-between mobile:items-start'>
                <SkeletonLoader $height='36px' $width='150px' $radius='10px' />
                <SkeletonLoader $height='36px' $width='190px' $radius='10px' />
                </div>
                <SkeletonLoader $height='36px' $width='80px' $radius='10px' />
            </div>
    </TileContainer>)
}

    return (
        <TileContainer className='p-10 gap-20 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23] mobile:flex-col mobile:gap-4'>
            <div className='flex flex-col justify-between'>
                <WalletAddressOverview address={wallet} className='mobile:hidden mb-[54px]' />
                <div className='mobile:border mobile:border-[#312F3A] mobile:rounded-[10px] mobile:p-4 mobile:w-full'>
                    <Body className='text-purple_light mb-2'>{t`title.totalBalance`}</Body>
                    <div className='flex gap-4 items-end'>
                        <H1 className='text-[50px]'>${total_wallet_balance.toFixed(2)}</H1>
                        <TokensPercentage percentage={percentageChange} />
                    </div>

                </div>
            </div>
            <div className='flex w-full flex-col items-end justify-end mobile:justify-normal gap-5'>
                <div className='flex w-full gap-4 justify-end mobile:flex-row tablet:flex-col tablet:items-end mobile:flex-row mobile:justify-between'>
                    <Tags icon={BlendIcon} tagText={`${numberOfTokens} ${t`label.tokens`}`} />
                    <Tags icon={RoutingIcon} tagText={`${t`helper.across`} ${numberOfBlockchains} ${numberOfBlockchains > 1 ? t`helper.chainSeveral` : t`helper.chainOne`}`} />
                </div>
                {allBlockchainsLogos.length && <TokensHorizontalList logos={allBlockchainsLogos || []} />}
            </div>
        </TileContainer>
    )
}

export default PortfolioOverview;
