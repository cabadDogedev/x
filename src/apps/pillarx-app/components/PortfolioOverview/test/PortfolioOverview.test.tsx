import renderer from 'react-test-renderer';

// components
import PortfolioOverview from '../PortfolioOverview';

// types
import { ApiLayout, Projection } from '../../../../../types/api';

describe('<PortfolioOverview />', () => {
    const mockData: Projection = {
        meta: {
            display: {
                title: 'title',
            },
        },
        data: {
            wallet: 'wallet_address',
            wallets: [''],
            total_wallet_balance: 1000,
            assets: [],
            total_realized_pnl: 53,
            total_unrealized_pnl: 12,
            total_pnl_history: { '24h': { realized: 100, unrealized: 200 } }
        },
        layout: ApiLayout.OVERVIEW,
        id: 'overview',        
    };
    const mockLoading = false;
  
    it('renders correctly and matches snapshot', () => {
        const tree = renderer
            .create(<PortfolioOverview data={mockData} isDataLoading={mockLoading} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('displays loading skeleton when data is loading', () => {
        const tree = renderer
            .create(<PortfolioOverview data={undefined} isDataLoading={true} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

});
