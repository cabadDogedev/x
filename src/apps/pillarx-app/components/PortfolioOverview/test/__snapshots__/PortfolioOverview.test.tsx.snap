// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PortfolioOverview /> displays loading skeleton when data is loading 1`] = `
.c0 {
  height: 40px;
  width: 200px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 54px;
}

.c1 {
  height: 77px;
  width: 180px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 0;
}

.c2 {
  height: 36px;
  width: 150px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 0;
}

.c3 {
  height: 36px;
  width: 190px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 0;
}

.c4 {
  height: 36px;
  width: 80px;
  animation: hLhBjP 1s linear infinite alternate;
  border-radius: 10px;
  margin-bottom: 0;
}

<div
  className="flex bg-container_grey rounded-2xl p-10 gap-20 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23] mobile:flex-col mobile:gap-4"
>
  <div
    className="flex flex-col justify-between"
  >
    <div
      className="c0 mobile:hidden"
    />
    <div
      className="mobile:border mobile:border-[#312F3A] mobile:rounded-[10px] mobile:p-4 mobile:w-full"
    >
      <div
        className="c1"
      />
    </div>
  </div>
  <div
    className="flex w-full flex-col items-end justify-end mobile:justify-normal gap-5"
  >
    <div
      className="flex w-full gap-4 justify-end mobile:flex-row tablet:flex-col tablet:items-end mobile:flex-row mobile:justify-between mobile:items-start"
    >
      <div
        className="c2"
      />
      <div
        className="c3"
      />
    </div>
    <div
      className="c4"
    />
  </div>
</div>
`;

exports[`<PortfolioOverview /> renders correctly and matches snapshot 1`] = `
<div
  className="flex bg-container_grey rounded-2xl p-10 gap-20 tablet:p-5 mobile:p-0 mobile:bg-[#1F1D23] mobile:flex-col mobile:gap-4"
>
  <div
    className="flex flex-col justify-between"
  >
    <div
      className="flex gap-2.5 items-center mobile:hidden mb-[54px]"
    >
      <div
        className="bg-[#312F3A] p-2.5 rounded-full w-10 h-10"
      >
        <img
          className="w-full h-full"
          src={
            {
              "ReactComponent": {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "default": "profile-icon.svg",
            }
          }
        />
      </div>
      <p
        className="text-base font-medium undefined"
      >
        wallet
        ...
        dress
      </p>
      <div
        className="w-4 h-4"
        onClick={[Function]}
      >
        <img
          className="cursor-pointer"
          src={
            {
              "ReactComponent": {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "default": "copy-icon.svg",
            }
          }
          style={
            {
              "height": "full",
              "width": "full",
            }
          }
        />
      </div>
    </div>
    <div
      className="mobile:border mobile:border-[#312F3A] mobile:rounded-[10px] mobile:p-4 mobile:w-full"
    >
      <p
        className="text-base font-medium text-purple_light mb-2"
      >
        title.totalBalance
      </p>
      <div
        className="flex gap-4 items-end"
      >
        <h1
          className="text-3xl leading-[45px] font-medium mobile:text-xl mobile:leading-5 text-[50px]"
        >
          $
          1000.00
        </h1>
        <div
          className="flex gap-1 items-center"
        >
          <img
            className="w-[10px] h[9px]"
            src={
              {
                "ReactComponent": {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "default": "triangle-up-green.svg",
              }
            }
          />
          <p
            className="text-sm font-medium text-percentage_green"
          >
            30.00
            %
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="flex w-full flex-col items-end justify-end mobile:justify-normal gap-5"
  >
    <div
      className="flex w-full gap-4 justify-end mobile:flex-row tablet:flex-col tablet:items-end mobile:flex-row mobile:justify-between"
    >
      <div
        className="flex gap-2 items-center bg-[#312F3A] py-1.5 px-4 rounded-md mobile:px-3"
      >
        <img
          src={
            {
              "ReactComponent": {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "default": "blend-icon.svg",
            }
          }
        />
        <p
          className="text-base font-medium undefined"
        >
          0 label.tokens
        </p>
      </div>
      <div
        className="flex gap-2 items-center bg-[#312F3A] py-1.5 px-4 rounded-md mobile:px-3"
      >
        <img
          src={
            {
              "ReactComponent": {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "default": "routing-icon.svg",
            }
          }
        />
        <p
          className="text-base font-medium undefined"
        >
          helper.across 0 helper.chainOne
        </p>
      </div>
    </div>
    0
  </div>
</div>
`;
